### PR TITLE
[SPARK-7633] [MLlib] [PySpark] Python bindings for StreamingLogisticRegressionwithSGD

### PR DESCRIPTION
**Pull request by MechCoder**
Add Python bindings to StreamingLogisticRegressionwithSGD.

No Java wrappers are needed as models are updated directly using train.

### CODE REVIEW

**Comment by MechCoder** [comment]
jenkins retest this please

**Comment by MechCoder** [comment]
cc @freeman-lab :P

**Comment by MechCoder** [comment]
@mengxr Could you review this? This should be the next one to go as I have refactored the code.

File: python/pyspark/mllib/classification.py [hunk 1]
```diff
@@ -24,11 +24,14 @@
 from pyspark.mllib.common import callMLlibFunc, _py2java, _java2py
 from pyspark.mllib.linalg import DenseVector, SparseVector, _convert_to_vector
 from pyspark.mllib.regression import LabeledPoint, LinearModel, _regression_train_wrapper
+from pyspark.streaming import DStream
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: organize imports

File: python/pyspark/mllib/classification.py [hunk 2]
```diff
@@ -24,11 +24,14 @@
...
 from pyspark.mllib.util import Saveable, Loader, inherit_doc
 
 
-__all__ = ['LogisticRegressionModel', 'LogisticRegressionWithSGD', 'LogisticRegressionWithLBFGS',
-           'SVMModel', 'SVMWithSGD', 'NaiveBayesModel', 'NaiveBayes']
+__all__ = ['LogisticRegressionModel', 'LogisticRegressionWithSGD',
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: No need to change this line.

File: python/pyspark/mllib/classification.py [hunk 3]
```diff
@@ -580,6 +583,102 @@ def train(cls, data, lambda_=1.0):
...
+class StreamingLinearAlgorithm(object):
+    """
+    Base class that has to be inherited by any StreamingLinearAlgorithm.
+
+    Prevents reimplementation of methods predictOn and predictOnValues.
+    """
+    def __init__(self, model):
+        self._model = model
+
+    def latestModel(self):
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: Should it be a `@property`? Maybe we should update `lastModel` in streaming k-means as well.

File: python/pyspark/mllib/classification.py [hunk 4]
```diff
@@ -580,6 +583,102 @@ def train(cls, data, lambda_=1.0):
...
+                "dstream should be a DStream object, got %s" % type(dstream))
+        if not self._model:
+            raise ValueError(
+                "Model must be intialized using setInitialWeights")
+
+    def predictOn(self, dstream):
+        """
+        Make predictions on a dstream.
+
+        Returns a transformed dstream object.
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: `:return:`

File: python/pyspark/mllib/classification.py [hunk 5]
```diff
@@ -580,6 +583,102 @@ def train(cls, data, lambda_=1.0):
...
+
+
+@inherit_doc
+class StreamingLogisticRegressionWithSGD(StreamingLinearAlgorithm):
+    """
+    Run LogisticRegression with SGD on a stream of data.
+
+    The weights obtained at the end of training a stream are used as initial
+    weights for the next stream.
+    :param stepSize: Step size for each iteration of gradient descent.
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: I think you need an empty line before params. Try `make html` under `python/docs` and check the generated html files.

File: python/pyspark/mllib/classification.py [hunk 6]
```diff
@@ -580,6 +583,102 @@ def train(cls, data, lambda_=1.0):
...
+    :param numIterations: Number of iterations run for each batch of data.
+    :param miniBatchFraction: Fraction of data on which SGD is run for each
+                              iteration.
+    :param regParam: L2 Regularization parameter.
+    """
+    def __init__(self, stepSize=0.1, numIterations=50,
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: Keep line width at 100.

File: python/pyspark/mllib/classification.py [hunk 7]
```diff
@@ -580,6 +583,102 @@ def train(cls, data, lambda_=1.0):
...
+        self._model = LogisticRegressionModel(
+            initialWeights, 0, initialWeights.size, 2)
+
+    def trainOn(self, dstream):
+        """Train the model on the incoming dstream."""
+        self._validate(dstream)
+
+        def update(rdd):
+            # Check for empty RDD Streams. If this try except clause is
+            # removed, then calling train raises an error.
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: Could you explain this issue?

File: python/pyspark/mllib/classification.py [hunk 8]
```diff
@@ -580,6 +583,102 @@ def train(cls, data, lambda_=1.0):
...
+            try:
```
Comment (mengxr) [inline python/pyspark/mllib/classification.py]: Shouldn't use `try ... except` to test this. Use `if rdd.isEmpty():` instead.

File: python/pyspark/mllib/tests.py [hunk 9]
```diff
@@ -25,7 +25,8 @@
 import array as pyarray
 from time import time, sleep
 
-from numpy import array, array_equal, zeros, inf, all, random
+from numpy import (
+    array, array_equal, zeros, inf, random, exp, dot, all, mean)
```
Comment (mengxr) [inline python/pyspark/mllib/tests.py]: Just curious, we don't need `(...)`, right? Btw, it is safer to write `import numpy as np` and then use `np.array` in the code. For example, `all` is a built-in method of Python. We can update this in a separate PR.

File: python/pyspark/mllib/tests.py [hunk 10]
```diff
@@ -1011,6 +1013,137 @@ def collect(rdd):
         self.assertEqual(predict_results, [[0, 1, 1], [1, 0, 1]])
 
 
+class StreamingLogisticRegressionWithSGDTests(MLLibStreamingTestCase):
+
+    @staticmethod
+    def generateLogisticInput(offset, scale, nPoints, seed):
```
Comment (mengxr) [inline python/pyspark/mllib/tests.py]: Please update the data generation to use #6715.

**Comment by mengxr** [comment]
@MechCoder I need to make another pass on the code and tests after your update. Please make sure the tests are the same as the ones on the Scala side.

File: python/pyspark/mllib/classification.py (hunk 3 above)
Comment (MechCoder) [inline python/pyspark/mllib/classification.py]: I followed the scala convention, where it is a method.

File: python/pyspark/mllib/classification.py (hunk 6 above)
Comment (MechCoder) [inline python/pyspark/mllib/classification.py]: Is it not 79? Or is the convention in PySpark 100?

File: python/pyspark/mllib/tests.py (hunk 9 above)
Comment (MechCoder) [inline python/pyspark/mllib/tests.py]: Sure, I thought this was some sort of micro-optimization and followed the convention that is already there.

File: python/pyspark/mllib/tests.py (hunk 10 above)
Comment (MechCoder) [inline python/pyspark/mllib/tests.py]: That is used to generate linear data with continuous labels.

**Comment by MechCoder** [comment]
@mengxr Thanks for the updates! I confirm that the tests are very similar

**Comment by MechCoder** [comment]
jenkins test this again

**Comment by MechCoder** [comment]
test this please

**Comment by MechCoder** [comment]
I have rerun the tests. (Strange that the tests that failed in the previous run were related to StreamingKMeans)

**Comment by mengxr** [comment]
@MechCoder The streaming k-means tests seem to be flaky. Could you run it multiple times on your local?

**Comment by MechCoder** [comment]
Sure, btw the last failure was some random error.

**Comment by MechCoder** [comment]
@mengxr Could you give another pass? I can update tomorrow, there is some problem with my ubunt.

**Comment by mengxr** [comment]
LGTM. Merged into master. I think the streaming k-means test might be flaky, which is usually hard to debug. Let's watch Jenkins master build: https://amplab.cs.berkeley.edu/jenkins/job/Spark-Master-SBT and see if there are failures.

**Pull request closed without merging**